In [5]:
## IMPORT PANDAS AND NUMPY.

import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
print("Libraries Loaded Successfully")

Libraries Loaded Successfully


In [7]:
## READ CSV FILE AND FIND NO. OF ROWS AND COLUMNS IN THE GIVEN DATASET

df = pd.read_csv("../data/raw/hotel_bookings.csv")
print(df.shape)

(119390, 32)


In [9]:
## GET INFORMATION OF THE DATASET (datatypes, count, entries, colums, rows etc.)

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

In [13]:
# DATA PREPROCESSING STARTS
## FIND MISSING VALUES/ENTRIES IN ROWS/COUMNS OF THE DATASET

missing_values = df.isnull().sum()

missing_values[missing_values > 0].sort_values(ascending=False)

company     112593
agent        16340
country        488
children         4
dtype: int64

In [12]:
## CALCULATE MISSING VALUES/ENTRIES PERCENTAGE IN ROWS/COUMNS OF THE DATASET

missing_percent = (df.isnull().sum() / len(df)) * 100

missing_percent[missing_percent > 0].sort_values(ascending=False)

company     94.306893
agent       13.686238
country      0.408744
children     0.003350
dtype: float64

## Initial Dataset Assessment

- Dataset contains 119,390 booking records.
- Dataset contains 32 features.
- Numerical and categorical variables are present.
- Several columns contain missing values.
- Company information appears largely unavailable and may require special treatment during preprocessing.

## Missing Value Analysis
### Findings

1. The `company` column contains 94.31% missing values.
   - Most bookings were not associated with a company.
   - This column may be removed during preprocessing.

2. The `agent` column contains 13.69% missing values.
   - Some bookings were made directly without an agent.
   - Missing values may be imputed or treated as a separate category.

3. The `country` column contains only 0.41% missing values.
   - Missing values can be filled using the mode.

4. The `children` column contains only 4 missing records.
   - These can be safely filled or removed.

In [15]:
## CHECK FOR DUPLICATE VALUES

duplicates = df.duplicated().sum()

print("Duplicated Rows :", duplicates )

Duplicated Rows : 31994


In [18]:
## CALCULATE PERCENTAGE OF DUPLICATE VALUES

duplicate_percentage = (duplicates / len(df)) * 100

print(f"Duplicate Percentage: {duplicate_percentage:.2f}%")

Duplicate Percentage: 26.80%


## Duplicate Record Analysis

- Dataset contains 31,994 duplicate records.
- Approximately 26.8% of the dataset consists of duplicates.
- Further investigation is required to determine whether these represent genuine repeated bookings or duplicated entries.
- Duplicate handling will be considered during data preprocessing.

In [16]:
## DEEP DESCRIBTION OF THE ENTERIES IN THE COLUMNS

df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000
mean,0.370416,104.011416,2016.156554,27.165173,15.798241,0.927599,2.500302,1.856403,0.103890,0.007949,0.031912,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363
std,0.482918,106.863097,0.707476,13.605138,8.780829,0.998613,1.908286,0.579261,0.398561,0.097436,0.175767,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000
25%,0.000000,18.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000
50%,0.000000,69.000000,2016.000000,28.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000
75%,1.000000,160.000000,2017.000000,38.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,1.000000,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000


In [23]:
df['is_canceled'].value_counts()


is_canceled
0    75166
1    44224
Name: count, dtype: int64

In [24]:
(df['is_canceled'].value_counts(normalize=True) * 100).round(2)

is_canceled
0    62.96
1    37.04
Name: proportion, dtype: float64

## Cancellation Analysis

### Findings

- Total bookings: 119,390
- Successfully completed bookings: 75,166 (62.96%)
- Canceled bookings: 44,224 (37.04%)

### Business Insight

The cancellation rate is significantly high, indicating a substantial impact on hotel revenue and occupancy planning.

### Machine Learning Perspective

The target variable is reasonably balanced, making the dataset suitable for classification modeling.

## Day 2 Summary
### Dataset Overview
- Rows: 119,390
- Columns: 32

### Data Types
- Integer columns: 16
- Float columns: 4
- Categorical/Text columns: 12

### Missing Values
- company: 94.31%
- agent: 13.69%
- country: 0.41%
- children: 0.003%

### Duplicate Records
- Duplicate rows: 31,994
- Duplicate percentage: 26.8%

### Target Variable
- Not canceled: 62.96%
- Canceled: 37.04%

### Initial Data Quality Concerns
- High missing values in company column
- Missing values in agent column
- Large number of duplicate records
- Potential outliers in ADR and guest counts